### PR TITLE
Comfortable Grid

### PR DESCRIPTION
**Pull request by jobobby04**
Add a comfortable grid (grid but with title under the image) setting in the library bottom sheet and as a setting in the switch mode action in browse source

### CODE REVIEW

File: app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferenceKeys.kt [hunk 1]
```diff
@@ -73,7 +73,7 @@ object PreferenceKeys {
 
     const val lastUsedCategory = "last_used_category"
 
-    const val catalogueAsList = "pref_display_catalogue_as_list"
+    const val catalogueViewSetting = "pref_display_catalogue_view_setting"
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferenceKeys.kt:76]: `catalogueDisplayMode`?

File: app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferenceKeys.kt [hunk 2]
```diff
@@ -131,7 +131,7 @@ object PreferenceKeys {
 
     const val downloadNewCategories = "download_new_categories"
 
-    const val libraryAsList = "pref_display_library_as_list"
+    const val libraryViewSetting = "pref_display_library_view_setting"
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferenceKeys.kt:134]: `libraryDisplayMode`?

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt [hunk 3]
```diff
@@ -187,7 +187,7 @@ open class BrowseSourceController(bundle: Bundle) :
             binding.catalogueView.removeView(oldRecycler)
         }
 
-        val recycler = if (presenter.isListMode) {
+        val recycler = if (presenter.mode == 1) {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt:190]: Consider adding an enum in `PreferenceValues` for these.

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt [hunk 4]
```diff
@@ -89,7 +89,7 @@ open class BrowseSourcePresenter(
     /**
      * Whether the view is in list mode or not.
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt:90]: Update this.

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt [hunk 5]
```diff
@@ -199,8 +199,8 @@ open class BrowseSourcePresenter(
      *
      * @param asList whether the current mode is in list or not.
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt:200]: Update this.

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt [hunk 6]
```diff
@@ -289,7 +289,13 @@ open class BrowseSourcePresenter(
...
     fun swapDisplayMode() {
-        prefs.catalogueAsList().set(!isListMode)
+        prefs.catalogueViewSetting().set(
+            when (mode) {
+                0 -> 1
+                1 -> 2
+                else -> 0
+            }
+        )
     }
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt:299]: I feel like it'd make more sense to make it an actual options list rather than cycling through them.

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Class used to hold the displayed data of a manga in the catalogue, like the cover or the title.
+ * All the elements from the layout file "item_source_grid" are available in this class.
+ *
+ * @param view the inflated view for this holder.
+ * @param adapter the adapter handling this holder.
+ * @constructor creates a new catalogue holder.
+ */
+class SourceComfortableGridHolder(private val view: View, private val adapter: FlexibleAdapter<*>) :
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt:23]: You could probably extend the existing holder?

File: app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryComfortableGridHolder.kt [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * Class used to hold the displayed data of a manga in the library, like the cover or the title.
+ * All the elements from the layout file "item_source_grid" are available in this class.
+ *
+ * @param view the inflated view for this holder.
+ * @param adapter the adapter handling this holder.
+ * @param listener a listener to react to single tap and long tap events.
+ * @constructor creates a new library holder.
+ */
+class LibraryComfortableGridHolder(
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryComfortableGridHolder.kt:28]: Same here, could probably extend the other existing one.

File: app/src/main/res/layout/source_comfortable_grid_item.xml [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:background="@drawable/library_item_selector"
+    android:padding="4dp">
+
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:id="@+id/outside"
```
Comment (arkon) [inline app/src/main/res/layout/source_comfortable_grid_item.xml:11]: Don't think this is used.

File: app/src/main/res/layout/source_comfortable_grid_item.xml [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:background="@drawable/rounded_rectangle">
```
Comment (arkon) [inline app/src/main/res/layout/source_comfortable_grid_item.xml:14]: Don't think this is needed either?

File: app/src/main/res/layout/source_comfortable_grid_item.xml [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                android:background="?attr/colorSurface"
+                tools:ignore="ContentDescription"
+                tools:src="@mipmap/ic_launcher" />
+
+            <View
+                android:id="@+id/gradient"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_gravity="bottom"
+                android:background="@drawable/gradient_shape" />
```
Comment (arkon) [inline app/src/main/res/layout/source_comfortable_grid_item.xml:38]: Not much use in there being a gradient scrim if there's no overlaid text.

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -83,6 +83,7 @@
     <string name="action_display">Display</string>
     <string name="action_display_grid">Grid</string>
```
Comment (arkon) [inline app/src/main/res/values/strings.xml:84]: Should probably reword this to be "Compact grid" to match J2K.

File: app/src/main/res/layout/source_comfortable_grid_item.xml (hunk 10 above)
Comment (jobobby04) [reply app/src/main/res/layout/source_comfortable_grid_item.xml:14]: This is used, there is a visible outline around the text, this rounds it(everything else you have said is correct)

Comment (arkon) [reply app/src/main/res/layout/source_comfortable_grid_item.xml:14]: Seems like an odd design choice. Do you have a screenshot?

Comment (jobobby04) [reply app/src/main/res/layout/source_comfortable_grid_item.xml:14]: I posted one on Discord just now

File: app/src/main/res/layout/source_comfortable_grid_item.xml (hunk 11 above)
Comment (jobobby04) [reply app/src/main/res/layout/source_comfortable_grid_item.xml:37]: I tried removing this, it did not go well

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt (hunk 6 above)
Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourcePresenter.kt]: Done, they are now a selection menu

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt (hunk 3 above)
Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt]: I added const values in the PreferenceValues, I am not sure how a enum would work(though this should be equivalent)

File: app/src/main/res/layout/source_comfortable_grid_item.xml (hunk 11 above)
Comment (arkon) [reply app/src/main/res/layout/source_comfortable_grid_item.xml:37]: Why?

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt (hunk 3 above)
Comment (arkon) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt]: I'd prefer an enum (and I should probably clean up the other ones). It's easier to see something like `DisplayMode.LIST` than `1`, or even separate consts like `DISPLAY_MODE_LIST`.

File: app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferencesHelper.kt [hunk 13]
```diff
@@ -143,7 +143,7 @@ class PreferencesHelper(val context: Context) {
 
     fun lastVersionCode() = flowPrefs.getInt("last_version_code", 0)
 
-    fun catalogueAsList() = rxPrefs.getBoolean(Keys.catalogueAsList, false)
+    fun catalogueDisplayMode() = flowPrefs.getInt(Keys.catalogueDisplayMode, 0)
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferencesHelper.kt:146]: Use the enum (or const, if you can't get that to work) everywhere.

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceItem.kt [hunk 14]
```diff
@@ -4,26 +4,28 @@ import android.view.Gravity
 import android.view.View
 import android.view.ViewGroup.LayoutParams.MATCH_PARENT
 import android.widget.FrameLayout
+import androidx.constraintlayout.widget.ConstraintLayout
 import androidx.recyclerview.widget.RecyclerView
-import com.f2prateek.rx.preferences.Preference
+import com.tfcporciuncula.flow.Preference
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceItem.kt:9]: Does this no longer crash?

File: app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryComfortableGridHolder.kt [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    adapter: FlexibleAdapter<IFlexible<RecyclerView.ViewHolder>>
+) : LibraryGridHolder(view, adapter) {
+
+    /**
+     * Method called from [LibraryCategoryAdapter.onBindViewHolder]. It updates the data for this
+     * holder with the given manga.
+     *
+     * @param item the manga item to bind.
+     */
+    override fun onSetValues(item: LibraryItem) {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryComfortableGridHolder.kt:39]: Isn't most or all of this the same as the parent function?

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        GlideApp.with(view.context).clear(thumbnail)
+        if (!manga.thumbnail_url.isNullOrEmpty()) {
+            GlideApp.with(view.context)
+                .load(manga.toMangaThumbnail())
+                .diskCacheStrategy(DiskCacheStrategy.DATA)
+                .centerCrop()
+                .placeholder(android.R.color.transparent)
+                .into(StateImageViewTarget(thumbnail, progress))
+        }
+    }
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt:55]: Isn't most or all of this the same as the parent function?

File: app/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -81,8 +81,9 @@
     <string name="action_migrate">Migrate</string>
     <string name="action_display_mode">Display mode</string>
     <string name="action_display">Display</string>
-    <string name="action_display_grid">Grid</string>
+    <string name="action_display_grid">Compact Grid</string>
```
Comment (arkon) [inline app/src/main/res/values/strings.xml:84]: Nitpick: lowercase for "grid"

File: app/src/main/res/menu/source_browse.xml [hunk 18]
```diff
@@ -10,11 +10,24 @@
...
+            <group android:checkableBehavior="single">
+                <item
+                    android:id="@+id/action_compact_grid"
+                    android:title="@string/action_display_grid" />
+                <item
+                    android:id="@+id/action_list"
+                    android:title="@string/action_display_list" />
+                <item
+                    android:id="@+id/action_comfortable_grid"
+                    android:title="@string/action_display_comfortable_grid" />
```
Comment (arkon) [inline app/src/main/res/menu/source_browse.xml:27]: Keep the grid options together.

File: app/src/main/res/layout/source_comfortable_grid_item.xml (hunk 11 above)
Comment (jobobby04) [reply app/src/main/res/layout/source_comfortable_grid_item.xml:37]: I'm not sure, but the result was that the manga cover turned super thin

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceItem.kt (hunk 14 above)
Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceItem.kt:9]: I removed the subscription because it was turned into a option, so I believe it should not crash anymore

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt (hunk 16 above)
Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt:55]: The imports are different, so I am not sure how it should be done

Comment (arkon) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt:55]: Oh right, since it's importing the synthetic layout view references? I guess it's fine for now then.

Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceComfortableGridHolder.kt:55]: Exactly

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceItem.kt (hunk 14 above)
Comment (arkon) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceItem.kt:9]: Got it, I'm going to do a quick change on my end first to remove all of the RxPrefs so that this PR won't involve that change.

Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/SourceItem.kt:9]: I rebased the PR on the latest 👍

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt (hunk 3 above)
Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/browse/BrowseSourceController.kt]: Its just I am not sure how to make a enum return a int for a option

File: app/src/main/res/menu/source_browse.xml (hunk 18 above)
Comment (jobobby04) [reply app/src/main/res/menu/source_browse.xml:24]: Done in latest commit

File: app/src/main/res/values/strings.xml (hunk 17 above)
Comment (jobobby04) [reply app/src/main/res/values/strings.xml]: Added

File: app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferencesHelper.kt (hunk 13 above)
Comment (jobobby04) [reply app/src/main/java/eu/kanade/tachiyomi/data/preference/PreferencesHelper.kt]: Added
